### ISSUE TITLE
Raspberry Pi 5: libm.so.6: ELF load command address/offset not page-aligned

### ISSUE THREAD

**Issue by sundhaug92** [ptitSeb/box86#912]
Raspberry Pi OS 64-bit

LSB
```
Distributor ID: Debian
Description:    Debian GNU/Linux 12 (bookworm)
Release:        12
Codename:       bookworm
```

uname: `Linux raspberrypi 6.1.0-rpi7-rpi-2712 #1 SMP PREEMPT Debian 1:6.1.63-1+rpt1 (2023-11-24) aarch64 GNU/Linux`

Happens when running test01, or steam

**Comment by ptitSeb** [comment]
You are probably using a 16K pagesize OS. Note that mostly nothing will work with 16K pagesize, at least not wine.

But you'll have to gives more detail if you are interrested in a fix, because I don't have a PI5 so I cannot reproduce anything on my side.

**Comment by sundhaug92** [comment]
> You are probably using a 16K pagesize OS. Note that mostly nothing will work with 16K pagesize, at least not wine.
> 
> But you'll have to gives more detail if you are interrested in a fix, because I don't have a PI5 so I cannot reproduce anything on my side.

Yeah after looking at the error for box64 I googled it and noticed that at least on rpi5 Raspberry OS is 16K pages

**Comment by StrikeNP** [comment]
I can second that I'm also running into this issue with the Raspberry Pi 5. It was also an issue with Box64 until I ran it as `DRPI5ARM64PS16K`, but it would appear that that's not an option on Box 86.

**Comment by StrikeNP** [comment]
I recompiled my kernel to 4K using the guide here:
https://www.raspberrypi.com/documentation/computers/linux_kernel.html#building-the-kernel-locally
While it did get my box86 past it's error, it did not resolve the issue documented here on box64 https://github.com/ptitSeb/box64/issues/91

**Comment by StrikeNP** [comment]
It would appear logging out and back in after running install_steam.sh helped me get past some issues, but now when I got to launch steam it just hangs.

**Comment by sundhaug92** [comment]
> I can second that I'm also running into this issue with the Raspberry Pi 5. It was also an issue with Box64 until I ran it as `DRPI5ARM64PS16K`, but it would appear that that's not an option on Box 86.
> 
> It would appear that either the RP5 kernel needs to be recompiled with 4K pages or box86 needs to add support for the RP5's 16K. I tried manually adding the compiler flags in the `CMakeLists.txt` file, but to no surprise it threw an error during compilation and failed. I've attached the compile log for anyone who's curious. [compile_log.txt](https://github.com/ptitSeb/box86/files/13693555/compile_log.txt)

Yeah I wrote the `RPI5ARM64(PS16K)` stuff after noticing this. One option is to for now use Ubuntu (which uses 4K pages on RPi5)

**Comment by theofficialgman** [comment]
@sundhaug92 kernel recompilation is not necessary on piOS. both 16k and 4k pagesize kernels are included, you can switch between the two with an option in config.txt https://github.com/raspberrypi/bookworm-feedback/issues/107#issuecomment-1773810662

**Comment by MichaIng** [comment]
Since there is no dedicated RPi 5 target yet, did you use `-DRPI4=1` or just `-DARM_DYNAREC=1`.

**Comment by IamSmeagol** [comment]
Hi I am having this issue also on a raspberry pi 5 and am not sure how to proceed
I tried installing based on https://github.com/ptitSeb/box86/blob/master/docs/COMPILE.md#for-raspberry-pi-on-64bit-os and also installing with https://github.com/ptitSeb/box86/blob/master/docs/COMPILE.md#for-raspberry-pi-on-64bit-os , but box86 always tells me `box86: error while loading shared libraries: libm.so.6: ELF load command address/offset not page-aligned`
What can I do to resolve this?

**Comment by MichaIng** [comment]
Again guys, there is no `-DRPI5ARM64=1`, so which target did you use?

The RPi 5 `*-2712` kernel has 16k page size, which makes it incompatible with 32-bit ARM binaries. I do not understand the underlying reasons well enough, but it might be a general blocker for 32-bit binaries (thus any multiarch binary packages, thus Box64).

RPi OS 64-bit also has the `*-v8` kernel pre-installed, which is compatible with all 64-bit RPi models, including RPi 5, and uses 4k page size. It can be selected explicitly via `config.txt`, or by uninstalling `linux-image-rpi-2712`. That should help, but `-DRPI4ARM64=1` strictly seen sets the wrong host CPU flags for the compiler: `cortex-a72` vs `cortex-a76`

**Comment by IamSmeagol** [comment]
Sorry I wasn't aware at first. I used `-DRPI4ARM64=1` and got the error above. 

I was able to run box86 by using `DARM_DYNAREC=1` instead in cmake. and adding `kernel=kernel8.img` to `/boot/firmware/config.txt` so that I have a 4k pagesize. Thank you

**Comment by ptitSeb** [comment]
On 64bits OS, I would suggest, again, to use box64 (with box32 to run 32bits apps), as it is also compatibile with 16k pagesize and is able to run many things there. Box86 is not compatible with 16k pagesize, and there is no plan to make it compatible.

**Comment by sundhaug92** [comment]
> Again guys, there is no `-DRPI5ARM64=1`, so which target did you use?

IIRC I added it to box64, not box86, and never got around to fixing the page-size stuff

**Comment by MichaIng** [comment]
I guess cortex-a76 optimization is minimal compared to cortex-a72, RPi OS 32-bit is rarely used on Raspberry Pi 5, now that the 64-bit variant is shown as first in the list on the download page as well as in RPi Imager, and one needs to manually prevent the 16k page size kernel from being used. But IMO it still makes sense to add the RPi 5 options, along with exactly that info, to avoid confusion:
- `-DRPI5` => for RPi OS 32-bit
- `-DRPI5ARM64` => for RPi OS 64-bit with 32-bit multiarch libs installed + 4k page size kernel used (`*-v8` instead of `*-2712` kernel package).

@ptitSeb 
That seems to cover a lot more/common RPi 5 use cases, indeed. Box32 is still shown as "experimental, do not use". Would you say it already better in practice than Box86 on 64-bit host with multiarch libs? And Box32 does not require multiarch libs, right?

**Comment by ptitSeb** [comment]
> [@ptitSeb](https://github.com/ptitSeb) That seems to cover a lot more/common RPi 5 use cases, indeed. Box32 is still shown as "experimental, do not use". Would you say it already better in practice than Box86 on 64-bit host with multiarch libs? And Box32 does not require multiarch libs, right?

Yeah, I will change the readme for box64 to make box32 not "do not use" anymore, just still "beta" software
